p2p/conn/connection.go

```diff
@@ -315,13 +337,21 @@ FOR_LOOP:
 			c.Logger.Debug("Send Ping")
 			wire.WriteByte(packetTypePing, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
+			c.Logger.Debug("Starting pong timer", "dur", c.config.PongTimeout)
+			c.pongTimer = time.AfterFunc(c.config.PongTimeout, func() {
+				c.pongTimeoutCh <- struct{}{}
+			})
 			c.flush()
+		case <-c.pongTimeoutCh:
+			c.Logger.Debug("Pong timeout")
+			err = errors.New("pong timeout")
 		case <-c.pong:
 			c.Logger.Debug("Send Pong")
 			wire.WriteByte(packetTypePong, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
 			c.flush()
 		case <-c.quit:
+			c.stopPongTimer()
```

Comment:
if theres an error in this loop, we just break out without processing the closed quit channel so pong timer won't get stopped. but we can't put it in OnStop because it will race. So I think it needs to also be called in the error != nil block below

Reply:
You're right. Thanks